### ISSUE TITLE
Add ArbitraryBasis and ArbitraryOrthogonalBasis

### ISSUE THREAD

**Issue by sethaxen** [JuliaManifolds/Manifolds.jl#106]
We should add `ArbitraryOrthogonalBasis` and `ArbitraryBasis`, with defaults

```julia
function get_vector(M::Manifold, x, v, B::ArbitraryBasis)
    return get_vector(M, x, v, ArbitraryOrthogonalBasis(number_system(B))
end
function get_vector(M::Manifold, x, v, B::ArbitraryOrthogonalBasis)
    return get_vector(M, x, v, ArbitraryOrthonormalBasis(number_system(B))
end

function get_coordinates(M::Manifold, x, v, B::ArbitraryBasis)
    return get_coordinates(M, x, v, ArbitraryOrthogonalBasis(number_system(B))
end
function get_coordinates(M::Manifold, x, v, B::ArbitraryOrthogonalBasis)
    return get_coordinates(M, x, v, ArbitraryOrthonormalBasis(number_system(B))
end

vee(M::Manifold, x, v) = get_coordinates(M, x, v, ArbitraryBasis())

hat(M::Manifold, x, v) = get_vector(M, x, v, ArbitraryBasis())
```

The `Euclidean` and `Rotations`, implementations of `vee!` and `hat!` will be switched to `get_coordinates`/`get_vector`. This changes `vee` and `hat` to be useful shorthands for "some vector representation of the vector space and its inverse" and standardizes our use of bases everywhere.

**Comment by kellertuer** [comment]
Why not `AbstractBasis`? Are there real implementations that make this generic basis non-abstract?

**Comment by kellertuer** [comment]
But I agree to the idea of changing `vee` and `hat`, that's a good idea.

**Comment by mateuszbaran** [comment]
I think we can do this.

> Why not `AbstractBasis`? Are there real implementations that make this generic basis non-abstract?

At the moment we have no non-orthonormal bases since I didn't know what they would be useful for. `ArbitraryBasis` would be the first. I don't understand though where would you want that `AbstractBasis`? `get_coordinates` and `get_vector` expect an object of a non-abstract type.

**Comment by kellertuer** [comment]
I was merely expecting that `ArbitraryBasis` is an `AbstractBasis` in some sense? What makes it different from being abstract?

**Comment by mateuszbaran** [comment]
Well, it's concrete, that is you can make an object of the type `ArbitraryBasis` but you can't make an object of type `AbstractBasis`. It's the same thing as in the case of `ArbitraryOrthonormalBasis`. Would you like to be able to call, for example, `get_vector(M, x, v, AbstractBasis)`? I think that would be confusing.

**Comment by kellertuer** [comment]
Ah, I wasn't aware that you could actually create that, sorry. Then I was just confused.

**Comment by mateuszbaran** [comment]
No problem 🙂 .

**Comment by kellertuer** [comment]
Since I was at basis anyways, I added these functions and noticed

1. I still don't like Arbitrary in here, it sounds a little strange. Unprecise maybe, where you would expect a type to be precise. Can we just call them Basis, `OrthogonalBasis` and `OrthonormalBasis`? Or maybe with a `Tangent` upfront?
2. The initial comment here proposes to introduce `hat` and `vee` which currentlyy default to `hat!` and `vee!` what about those?

In general I like this idea of hat and vee getting general fallbacks :)

**Comment by mateuszbaran** [comment]
> 1. I still don't like Arbitrary in here, it sounds a little strange. Unprecise maybe, where you would expect a type to be precise. Can we just call them Basis, `OrthogonalBasis` and `OrthonormalBasis`? Or maybe with a `Tangent` upfront?

We can use a different word perhaps, `AnyBasis`, `FastBasis` or something. I'm not sure about just `Basis`, the name sounds vague and unspecific.

> 2\. The initial comment here proposes to introduce `hat` and `vee` which currentlyy default to `hat!` and `vee!` what about those?

I don't know, we could perhaps use `get_vector` and `get_coordinates` as a fallback for mutating `hat!` and `vee!`?

**Comment by kellertuer** [comment]
I think even `Any` is a little vague and either `Fast` or (as otherwise) `Default´ would be nice. Maybe really `Default`.

The reason I stubled was, because neither `get_vector` nor `get_coordinates` aren't mutating.

**Comment by mateuszbaran** [comment]
> I think even `Any` is a little vague and either `Fast` or (as otherwise) `Default´ would be nice. Maybe really `Default`.

OK, `Default` sounds fine.

> The reason I stumbled was, because neither `get_vector` nor `get_coordinates` aren't mutating.

We could add mutating variants of these two actually but it doesn't look particularly important. There would very rarely be significant performance gains by using mutating variants of these functions.

**Comment by kellertuer** [comment]
I renamed it to `Default`. I tried switching as the initial comment said, to `get_vector` and `get_coodrinates` and noticed that the `log!` pithing Rotations needed a mutating `get_vector`, so I implemented that – now it's just a few errors remaining that I again just don't see why they appeared, so I first have to read more about the first rotations and their closed forms to see where I might have mistaken the dimensions. Somewhere – though I hoped to only rename functions – a formula must have changed.

I hope though, the renaming was the one intended in this issue.
